### ISSUE TITLE
Avoid dot notation for setting configuration

### ISSUE THREAD

**Issue by yandodov** [getkirby/kirby#2697]
This issue builds upon getkirby/kirby#2343 and [a discussion we had in Slack](https://getkirby.slack.com/archives/C2UTAQAN4/p1594207467116700). [The docs](https://getkirby.com/docs/reference/plugins/extensions/options#registering-options) were updated to reflect the new changes to dot notation in the new [3.4.0 release](https://github.com/getkirby/kirby/releases/tag/3.4.0).

As I said in the mentioned issue, I think dot notation is absolutely useful, but only for _getting_ values, not for setting. I still think that's the case and here are my arguments.

# Problems

## 1. Possible misuse

The updated docs say (emphasis mine):

>**Always** register your options with the dot notation (like in the example above), **never as nested arrays** (meaning: each separate option should be defined on the top-level of the options array).

And in Slack, @lukasbestle gave the example of a plugin option with key `image.jpg`. Since users are encouraged to **always** use dot notation, you could end up with this:

```php
'options' => [
    'foo.image.jpg' => 'bar'
]
```

...which will be expanded to:

```php
'foo' => [
    'image' => [
        'jpg' => 'bar'
    ]
]
```

...and not the expected:

```php
'foo' => [
    'image.jpg' => 'baz'
]
```

## 2. Mental overhead

We can specify options as nested arrays in `config.php` and in the `App` constructor, but not in plugins. What about the `ready` callback, though? In a `ready` callback in `config.php` - with nested arrays and in `ready` callback of a plugin - with dot notation? Or we can use both? Or always with dot notation? Then, quoting the docs, there's this:

>Please also note that it's not possible to define both a plugin option without a dot and a plugin option that starts with the same name but then continues with a dot.

There are all those things you have to keep in mind for no apparent reason. It goes against the simplicity and intuitiveness of Kirby.

## 3. It's not a pattern

Where have you seen this pattern in other software? Have you ever seen something like this:

```js
let thing = new Thing({
    foo: 'bar',
    'my.nested.key': {
        'one': 1,
        'two': 2
    }
})
```

...in the JavaScript world, for example? Even the PHP one, for that matter. I haven't seen it and my guess for the reason why is ambiguity. You can do the same things in so many ways:

```php
'my' => [
    'nested' => [
        'key' => 42
    ]
]
```

```php
'my' => [
    'nested.key' => 42
]
```

```php
'my.nested' => [
    'key' => 42
]
```

```php
'my.nested.key' => 42
```

...and it only gets worse as you add nesting.

_What's the point of this?_ What issues does this solve? The way I see it, it only accomplished:

- bugs in plugins
- confusion among developers
- more complex Kirby core, and from what I know, @bastianallgeier likes to keep things lean
- this huge warning that users have to keep in mind ![image](https://user-images.githubusercontent.com/5570098/87026198-152a0100-c1e4-11ea-99b0-7e0344d5b61e.png)

Personally, this feature hasn't helped me in any way. At its best, it would save a couple lines of code. But given all the things I mentioned above, I hardly believe it's worth it. Maybe that's the reason it's not a widespread pattern.

# Solution

Kirby should **never** expand keys with dots. It should treat config as-is. You pass this:

```php
'options' => [
    'my.foo' => [
        'bar' => 1,
        'baz.qux' => 2
    ]
]
```

...and you get these options:

```php
'my.foo' => [
    'bar' => 1,
    'baz.qux' => 2
]
```

This way:

- You wouldn't need to remember warnings
- There's **one** way and **one way only** to set some option. Plugin expects an array, you give it an array, end of story. `config.php`? Yep, expects an array. `App` constructor? Still an array
- Kirby could reduce its options logic to just `array_merge_recursive()`. No need to process the input and expand keys, but then worrying that `image.jpg` shouldn't be expanded. If the user wants to have `image.jpg`, they would simply put that key in their source code...

Pretty simple and intuitive, right? The only downside is backwards compatibility - a lot of stuff depends on dot notation. This is why Kirby should somehow mark this as deprecated, with the intention of removing it _later_. In the meantime, the docs could start recommending using regular arrays always, everywhere, and replace the huge warning with one that says "setting configuration with dot notation is deprecated."

**Please note** - I'm against _setting_ configuration values with dot notation, not _accessing_ them. That's because while there are many different ways to assign a value (as I showed earlier), there is only one way to retrieve it.

---

Basically, all this options logic should be reduced to simply merging the option arrays from the `App` constructor, `config.php`, and plugins. What is known as "dot notation" should exist only in the `$kirby->option()` method as a way of doing this:

```php
option('foo.bar.baz.qux')
```

...rather than this:

```php
option('foo')['bar']['baz']['qux']
```

This way, dot notation would be just a tiny solution for a tiny problem, as it should be.

**Comment by distantnative** [comment]
I moved this to the kirby repo.
Since this has been a quite big breaking change in 3.4, I think this is a crucial discussion to be settled now and not just an idea for some point in the future.

**Comment by distantnative** [comment]
And I have to agree that having such a huge warning explanation in our docs is a clear sign that our current implementation might be not ideal.

**Comment by lukasbestle** [comment]
@hdodov Thanks for your detailed feedback.

A lot of thoughts went into the new implementation in 3.4.0 (you can take my word for that – I thought about a lot of edge cases). In this comment I want to explain some of my thoughts.

First, something I have fixed based on your feedback:

> And in Slack, @lukasbestle gave the example of a plugin option with key image.jpg. Since users are encouraged to always use dot notation, you could end up with this:

I don't expect many users to do this because the syntax from your example looks very weird and strange, but I have [updated the docs](https://github.com/getkirby/getkirby.com/commit/ea8175f3f7d481c4cfeacf7e3c43ba4d97ef7dd6) with an explanation that is a bit more clear.

---

Now let's take a look at your arguments one by one:

> We can specify options as nested arrays in config.php and in the App constructor, but not in plugins. What about the ready callback, though? In a ready callback in config.php - with nested arrays and in ready callback of a plugin - with dot notation? Or we can use both? Or always with dot notation?

The docs only talk about registering options, not setting them – after all it's the docs page on the options *plugin extension*. The `ready` callback (like the site options or the `App` constructor) sets options, so it will totally accept any variation of the two notations.

To be honest, I don't see how this will create mental overhead as users who haven't thought about the implementation details won't even question whether the syntax works in specific places or not. Unless they create plugins with plugin options (in which case they will see the warning paragraph and it will apply to them), they can just set options however they like and it will just work.

Which means: The only place that requires the dot notation are plugin option defaults. All other places can use whichever syntax you prefer and even a combination at any level of the tree.

Note that the only ones who have to deal with the warning are plugin devs – and they only need to deal with it once by deciding what their plugin options actually are. After that, it will all magically work as expected without having to think about this at all.

---

>> Please also note that it's not possible to define both a plugin option without a dot and a plugin option that starts with the same name but then continues with a dot.

> There are all those things you have to keep in mind for no apparent reason. It goes against the simplicity and intuitiveness of Kirby.

That's not new about the 3.4 implementation though and will also apply to an implementation that only accepts the dot notation for *getting* values. If the option getter fully supports the dot notation, it needs to query the different parts of the option name in an (internal) tree structure. And a tree that has two different nodes with the same name (one with its own value and one as the parent of other nodes) is just plain invalid and won't work.

The only reason why it has sometimes worked by accident before 3.4 is that the old implementation didn't normalize the tree, so some calls to the getter have seemingly worked as expected while others will have returned a completely invalid result or none at all. I have an example for this below.

---

> What's the point of this? What issues does this solve?
> Personally, this feature hasn't helped me in any way. At its best, it would save a couple lines of code. But given all the things I mentioned above, I hardly believe it's worth it. Maybe that's the reason it's not a widespread pattern.

The main advantages are:

**1. Makes it possible to set options in a nested structure in the config**

As site config files are just plain PHP files that return an array, it is possible to include other PHP files from them (for example to include protected data such as keys or passwords or to structure a long config by separating it into multiple files).

With 3.4, it has become possible to extract a whole set of options into an array and store it in a different file. For example:

```php
return [
  'superwoman.superplugin.foo' => [...],
  'superwoman.superplugin.bar' => 'bar',
  'superwoman.anotherplugin.foo' => [...]
];
```

can become:

```php
$foo = include 'foo.php';

return [
  'superwoman' => [
    'superplugin' => [
      'foo' => $foo,
      'bar' => 'bar'
    ],
    'anotherplugin.foo' => $foo
  ]
];
```

Any variation is possible, which makes the whole structure very flexible and easy to understand (you don't need to check whether the plugin actually uses the option structure you plan on using). It just works.™️

**2. Improved reliability of getting options**

Meaning: No matter how the option defaults are defined by the plugin dev and the options are set by the user, the plugin can get the values in a uniform way (see #2343, which was the reason why we changed the implementation in the first place).

**3. Getting a tree of options will return an array**

If the plugin has an option `my.option`, then the call to `$kirby->option('superwoman.superplugin.my')` will return an array with a key `option`. This on its own is not really useful, but imagine you have multiple options inside the same namespace (here: `my`) and you need to pass all of them to a method – easy peasy with 3.4, manual work before.

**4. Defining options that way as a best practice**

Another advantage of registering all plugin options on the top-level using the dot notation is that all separate options are clearly visible in the plugin code (I use this for documentation in the plugins I build for clients). In a nested structure, it's sometimes not clear if there is an option that is of type array or if the array values are the actual options.

Defining separate options separately makes it a lot more clear what the config structure looks like.

---

> Please note - I'm against setting configuration values with dot notation, not accessing them. That's because while there are many different ways to assign a value (as I showed earlier), there is only one way to retrieve it.

Unfortunately no, that's exactly the issue we solved in 3.4. Before 3.4, we had #2343, which was caused by exactly the misconception that there is only one way of getting values using the dot notation from a tree that is partly nested and that partly uses dot notation.

Imagine the following set of options:

```php
'options' => [
  'foo.bar' => [
    'baz' => 'foobar'
  ],
  'foo' => [
    'bar.baz' => 'foo'
  ]
]
```

These are the results for different dot notation queries (assuming that the options are not modified while setting as in 3.3.6 and prior releases):

```
foo         => ['bar.baz' => 'foo']
foo.bar     => ['baz' => 'foobar']
foo.bar.baz => Who knows? 🤯
```

The expected result would be:

```
foo         => ['bar' => ['baz' => 'foo']]
foo.bar     => ['baz' => 'foo']
foo.bar.baz => 'foo'
```

And that's not a theoretical example. That stuff has happened in real life and it's incredibly hard to debug. The new implementation in 3.4 on the other hand "just" works as you would expect, no matter which options are set from where in which nesting variation.

---

So to summarize: Of course we could say "we don't care about how options are set at all and let plugin devs decide the exact way their options need to be set in the config". However we have seen in #2343 that people will try to use a notation they have seen from other plugins with a plugin that actually uses a different option structure. That won't work and this is exactly what we have fixed in 3.4.0.

The new implementation in 3.4.0 is a lot more solid than the one before and really the only ones who need to be careful are the plugin devs. But as designing the structure of plugin options has always been a task that shouldn't be taken lightly (the structure has a huge effect on future expansions and on the ease of use for the plugin users), I think this is a place where it makes a lot of sense to think twice and ~cut~ code once.

My suggestion is: Let's wait for feedback from other plugin devs. @bnomei and @sebastiangreger have already replied with positive feedback in our Slack discussion, so it looks like the new implementation isn't impossible to use for plugin devs. And when comparing this minor additional task for plugin devs with the major reliability and flexibility improvement for users, I firmly believe that the rewrite in 3.4 was worth it.

**Comment by yandodov** [comment]
@lukasbestle my responses:

>The docs only talk about registering options, not setting them – after all it's the docs page on the options plugin extension.

Isn't registering basically _setting_ an initial value? I think the line between setting and registering is very subtle. If you're not familiar with the internals it can be unexpected that both have different behaviors.

>To be honest, I don't see how this will create mental overhead as users who haven't thought about the implementation details won't even question whether the syntax works in specific places or not. **Unless they create plugins with plugin options (in which case they will see the warning paragraph and it will apply to them)**

That's what I'm talking about. If you're a developer, you have to see the warning **and then remember it**. Sure, I'll see the warning. But what about 1 month later when I'm working a different plugin? Because there will be no fatal error for _not_ using dot notation, I'll probably get an issue about an unexpected behavior some time in the future which I'll have to debug.

I've had issues with options on several occasions - because I keep forgetting that options have a tricky behavior. This new constraint of using only dot notation in plugins is essentially the same thing.

>Note that the only ones who have to deal with the warning are plugin devs

So it's fine to make the job of plugin devs harder, as long as end users are not affected? I'm not sure that's part of the Kirby ethos. Wouldn't it be better if no one has to deal with anything?

>After that, it will all magically work as expected without having to think about this at all.

I'm not a fan of software that works magically. It's fine when it works _as expected_, but once it inevitably doesn't - you're on a magical journey of finding out how all this magic happens and figuring out an even more magical way around it.

>With 3.4, it has become possible to extract a whole set of options into an array and store it in a different file.

You don't need dot notation for that. You can put your nested array of options in a file and then include it wherever you want. Furthermore, if you have _two_ files, you can include them and merge them with `array_merge_recursive()`. If you use dot notation - you can't, because the options array is flattened.

>No matter how the option defaults are defined by the plugin dev and the options are set by the user, the plugin can get the values in a uniform way (see #2343, which was the reason why we changed the implementation in the first place).

There are two ways to set options, which has become a problem for getting them. You've changed the implementation to fix that. But that problem wouldn't have existed in the first place if there weren't two ways to set options? Maybe that's an easier fix?

>3. Getting a tree of options will return an array

Yes, and if you define these options as arrays in the first place, it would still return an array. The difference is that with arrays - it's more obvious, which is a good thing.

```php
'options' => [
    'foo.bar' => 42
]
```

Here, I've never defined the array `foo`, yet I expect to get it when using `option('vendor.plugin.foo')`. Kirby has to do that job for me and, more importantly, it rely on it to do it the way I want.

```php
'options' => [
    'foo' => [
        'bar' => 42
    ]
]
```

Here, `foo` is an array because it has square brackets and all. I defined it. When I call `option('vendor.plugin.foo')`, I expect that same array. This is still easy-peasy with the added benefit of being easier to understand.

>Another advantage of registering all plugin options on the top-level using the dot notation is that all separate options are clearly visible in the plugin code

Personally, I like this:

```php
'options' => [
    'nest' => [
        'foo' => 1, // this does foo
        'bar' => 2, // this does bar
        'baz' => 3  // this does baz
    ]
]
```

...more than this:

```php
'options' => [
    'nest.foo' => 1, // this does foo
    'nast.bar' => 2, // this does bar
    'nest.baz' => 3  // this does baz
]
```

...but that's up to preference. However, some notable differences are:

1. You have to write out the array name multiple times
2. You can misspell it and have a hard time figuring it out
3. If you want to change the name, you have to change it in multiple places

>Imagine the following set of options:
>
>```php
>'options' => [
>  'foo.bar' => [
>    'baz' => 'foobar'
>  ],
>  'foo' => [
>    'bar.baz' => 'foo'
>  ]
>]
>```
>These are the results for different dot notation queries [...] And that's not a theoretical example. That stuff has happened in real life and it's incredibly hard to debug.

That's a prime example of a problem that wouldn't have existed and be hard to debug if dot notation didn't exist in the first place. Why? Simply because you can't set the same value of an array to two different values.

```php
'options' => [
    'foo' => [
        'bar' => [
            'baz' => 'foo' // can't be `foobar` simultaneously
        ]
    ]
]
```

>The new implementation in 3.4.0 is a lot more solid than the one before and really the only ones who need to be careful are the plugin devs. But as designing the structure of plugin options has always been a task that shouldn't be taken lightly (the structure has a huge effect on future expansions and on the ease of use for the plugin users)

An even more solid solution is to not have any implementation at all and not have people be careful. And exactly because this shouldn't be taken lightly, it should be dead simple and intuitive. I think it can't get more simpler than simply merging arrays. You just need to look up the documentation of `array_replace_recursive()` (which Kirby currently uses for options anyways) and that's it. Kirby doesn't need to do anything special, users (including plugin devs) don't have to read and remember anything. They just need to know PHP.

**Comment by lukasbestle** [comment]
I don't agree with everything you write, but you are definitely right that the current implementation isn't perfect.

Because this is a very complex problem to solve, I need some time to think about different approaches. I will get back to you next week. :)

**Comment by gagarine** [comment]
I read this thread carefully and can see a good reason for the current implementation.

Use PHP array or an already existing format such as YAML or TOML and keep options in an external file (ie options.yml).

**Comment by lukasbestle** [comment]
Based on our discussion, I have thought about various ways forward and have come to the conclusion that the automatic array nesting in 3.4.0 needs to be removed/reverted again to keep the option feature easy to understand. That's also sad because of the major advantages the new implementation has, but overall I agree with @hdodov that it's more important to keep the implementation simple than to cover edge-cases in site config that can be avoided with a clear options syntax and documentation.

But just reverting the changes in 3.4.0 would being us back to the situation before 3.4.0 where people tried to use dot notation in places where it wasn't supported. Instead of adding support for dot notation in another way (which would have the same issues as the 3.4.0 implementation), my proposal is the following new syntax for plugin options in the `config.php`:

```php
<?php

return [
  'superwoman.superplugin' => [
    'superoption1' => 'value',
    'superoption2' => [
      'nested'
    ]
  ]
];
```

This makes it clear that everything inside the main plugin array needs to be set exactly like the plugin dev has documented in the plugin docs – unlike the current syntax of `superwoman.superplugin.superoption` (which makes people think that dot notation can also be used after the first two dots).

To ensure backwards-compatibility, options defined using the old syntax in the site config will be converted automatically to the new syntax when Kirby is loaded.

I have implemented this in #2739. Please take a look and test this with your real-world sites and plugins so we can get a feeling if this covers all use-cases and edge-cases. Option handling is a very complex feature, so it's important to do real-world testing.

Unfortunately there is one issue: Plugin devs who have already updated their plugins to define option defaults using the dot notation will have to revert their plugins to the old nested option structure as there will be no longer be the automatic nesting from 3.4.0.

I am looking forward to your opinion on this.

---

@hdodov One note regarding one argument in your latest comment:

> ```
> 'options' => [
>     'foo.bar' => 42
> ]
> ```
> 
> Here, I've never defined the array foo, yet I expect to get it when using option('vendor.plugin.foo'). Kirby has to do that job for me and, more importantly, it rely on it to do it the way I want.

This won't work when support for setting options with the dot notation is removed again like you propose and like I have done in my PR. If Kirby doesn't interpret the dots, it cannot split keys by dot to return the array you want.

As I explained in my previous comments, this was one of the (smaller) advantages of the implementation in 3.4.0, but given the complications that implementation has caused, I think it's OK to remove this feature again to make the option handling more robust and easier to understand overall.

**Comment by yandodov** [comment]
@lukasbestle 

>This makes it clear that everything inside the main plugin array needs to be set exactly like the plugin dev has documented in the plugin docs – unlike the current syntax of `superwoman.superplugin.superoption` (which makes people think that dot notation can also be used after the first two dots).

Actually, it still might hint that dot notation can be used due to the presence of a dot in `superwoman.superplugin`. It's still a familiar syntax and people who haven't paid enough attention to the changelog might get surprised. I think it would be clearer if you specify the plugin ID in the config. For example, I have this plugin:

```php
App::plugin('superwoman/superplugin', [
    'options' => [
        'foo' => 42
    ]
]);
```

...and I specify options like this:

```php
return [
    'superwoman/superplugin' => [
        'foo' => 64
    ]
];
```

I think there's less ambiguity this way because there are no dots involved. It's an option syntax that we've never had before and it will be harder to misinterpret it. Besides, Kirby already uses slashes to register the plugin `App::plugin('vendor/plugin')` instead of `App::plugin('vendor.plugin')` so it will also bring consistency and therefore simplicity.

Also, if I have:

```php
option('foo.bar.baz')
```

...how do I know if I'm getting the `baz` option of a `foo/bar` plugin, or if I'm just getting `baz` within `bar` within `foo`? You can argue it's a bit of an edge case, but it can still be confusing for less experienced people. With slashes, that's harder to misunderstand:

```php
option('foo/bar.baz')
```

**Edit:** I feel like the confusion comes from the fact that the dot `.` character plays two roles at once - defining a vendor namespace and retrieving a nested key. You've already established that slashes denote the namespace in the `App::plugin()` method, it only makes sense to continue with that pattern. I'm sure this can also simplify the logic [here](https://github.com/getkirby/kirby/pull/2739/files#diff-87fabdc89ce57d98f38c9307c1356a87R231), for example, since you should be able to just explode the string by `.`

---

>This won't work when support for setting options with the dot notation is removed again like you propose and like I have done in my PR. If Kirby doesn't interpret the dots, it cannot split keys by dot to return the array you want.

I actually used that as an argument _against_ dot notation. If I do `vendor.plugin.foo`, I am expecting to get an array that I never defined (Kirby did that for me). It's more straightforward to just define the array yourself. That way, you can only get what you define, which sounds about right.

---

@lukasbestle as a side question, are you removing nesting just for plugin options, or you've decided to slowly move away from dot notation setters altogether?

**Comment by lukasbestle** [comment]
> I think there's less ambiguity this way because there are no dots involved. It's an option syntax that we've never had before and it will be harder to misinterpret it. Besides, Kirby already uses slashes to register the plugin `App::plugin('vendor/plugin')` instead of `App::plugin('vendor.plugin')` so it will also bring consistency and therefore simplicity.

I agree, however this would be a breaking-change as all plugins would have to be updated to use the slash in their option getters. So I don't think it's viable to change this now.

> I'm sure this can also simplify the logic [here](https://github.com/getkirby/kirby/pull/2739/files#diff-87fabdc89ce57d98f38c9307c1356a87R231), for example, since you should be able to just explode the string by `.`

No, because the point of this code you linked to is to keep backwards-compatibility with the *old* syntax, which used `vendor.plugin.option`. Therefore the code cannot be simplified by changing the *new* syntax.

> I actually used that as an argument _against_ dot notation. If I do `vendor.plugin.foo`, I am expecting to get an array that I never defined (Kirby did that for me). It's more straightforward to just define the array yourself. That way, you can only get what you define, which sounds about right.

Ah, I misunderstood your argument. What you mean is that you actually *don't* expect Kirby to return an array here, right?

> as a side question, are you removing nesting just for plugin options, or you've decided to slowly move away from dot notation setters altogether?

Dot notation for setting options wasn't a thing before 3.4.0. By reverting that feature completely, we are getting rid of dot notation for setting options entirely in one go. This makes 3.4.1 an update with a breaking-change (which is something we don't generally do), but it's our only chance to revert this before more people actually use and rely on the feature we introduced in 3.4.0.

**Comment by distantnative** [comment]
All in all sounds very good (if I understand it correctly - setting = no dots, getting = use dots for getting array items).

If the dot in the plug-in author + plug-in name combination is causing irritation, would it be an option to also remove that part? And turn it into array and subarray as well?

**Comment by lukasbestle** [comment]
That would be an option, yes! But I think it would be a bit too verbose for everyday use as you would need two levels of nesting even if you only have one plugin from that plugin dev installed.

**Comment by yandodov** [comment]
@lukasbestle 

>What you mean is that you actually don't expect Kirby to return an array here, right?

Well, yes. It's not _clear_ what to expect. On one hand, I have a nested property, which can only exist in an array. On the other hand, I didn't define that array. You can expect two things to be returned - an array because the nested option needs a parent, or a falsy value because you haven't defined the parent option. Explicitly defining everything makes that clear.

>Dot notation for setting options wasn't a thing before 3.4.0.

Hm, I had the impression that it existed. I thought I'd seen it around the docs, but after looking around, I didn't find an example of it except for `remote.agent` [here](https://getkirby.com/docs/reference/system/options/remote#example) but that's from 3.4.0 which introduced the feature. Looks like I had a false impression.

@distantnative 

>If the dot in the plug-in author + plug-in name combination is causing irritation, would it be an option to also remove that part? And turn it into array and subarray as well?

I agree with Lukas that it can be too verbose. Which is why I think the slash solution is a good way to go. It would feel natural to define plugin options with the plugin ID. It would be easier to grasp from a newcomer's perspective: plugins must have an ID, you register them with that ID, you configure them with that ID, you use their options with that ID. Simple.

However, as Lukas said, that would be a breaking change. Is it an option have both syntaxes and mark the current one as deprecated? This would give users and plugin devs some time to react before you remove it in 3.5.0 or 3.6.0?

**Comment by lukasbestle** [comment]
> Well, yes. It's not _clear_ what to expect. On one hand, I have a nested property, which can only exist in an array. On the other hand, I didn't define that array. You can expect two things to be returned - an array because the nested option needs a parent, or a falsy value because you haven't defined the parent option. Explicitly defining everything makes that clear.

So to summarize: You are fine with the implementation from my PR?

> Is it an option have both syntaxes and mark the current one as deprecated? This would give users and plugin devs some time to react before you remove it in 3.5.0 or 3.6.0?

No, unfortunately that's not possible either, otherwise plugin devs would have to use *both* syntaxes for getting options, which would make everything even more complex.

**Comment by yandodov** [comment]
@lukasbestle 

>So to summarize: You are fine with the implementation from my PR?

Yep, yep! 👍

>No, unfortunately that's not possible either, otherwise plugin devs would have to use both syntaxes for getting options, which would make everything even more complex.

Yes, it will make it more complex until that later 3.5.0 or 3.6.0 release, but that's just for backwards compatibility and to avoid having a breaking change that catches everyone off-guard.

While there will be two syntaxes, plugin devs _should_ use _one_. The other one should be deprecated and discouraged. Does Kirby have a way to catch such warnings and show them to the user? For example, when you're in `debug` mode and you open the panel - a warning appears that says "plugin X uses the deprecated feature Y." As long as devs know that one of the syntaxes is outdated and that's clearly communicated, I don't see a big issue?

**Comment by lukasbestle** [comment]
As I wrote, plugin devs would need to use both syntaxes in the meantime. This would be a breaking-change.

It is not possible to convert between the syntaxes (slash and dot) without a lot of added complexity.

So this is unfortunately not viable.

**Comment by shoesforindustry** [comment]
Whilst I understand the reason for the change, I would also like a reversion to the older model. It caught me out today whilst upgrading: [See: $kirby->option() in 3.4.0 splits item when it has a dot in text?](https://forum.getkirby.com/t/kirby-option-in-3-4-0-splits-item-when-it-has-a-dot-in-text/18908/4)

We use this in one of our plugins and whilst I guess we could work around this, but the config seems the perfect place for this info.

**Comment by bastianallgeier** [comment]
✅  We've just merged the PR to reverse this change. It would be fantastic if you could test it during the 3.4.1-rc.1 starting today.

**Comment by sylvainjule** [comment]
Apologies for coming here long after the battle, @lukasbestle @bastianallgeier if I understand correctly this change should now be allowing both:

```php
'community.markdown-field.invisibles' => false,
// and
'community.markdown-field' => [
    'invisibles' => false,
]
```

Or do we need to convert everything to:

```php
'community.markdown-field' => [
    'key' => 'value',
    ...
]
```

Because both work when my options are booleans, but there seems to be an issue when using arrays:

```php
// doesn't work
'community.markdown-field.buttons'    => ['headlines', 'bold', 'image', 'file', 'pagelink', 'footnote'],
// works
'community.markdown-field'    => [
    'buttons' => ['headlines', 'bold', 'image', 'file', 'pagelink', 'footnote'],
]
```

Do you think we can have it working again or should every array option be now written as above? It breaks all my config files.

**Comment by lukasbestle** [comment]
@sylvainjule Sorry for the super late response, unfortunately we missed your comment and I now stumbled across it. 😢

You are right that both syntax variants should now be supported by Kirby. I tried to reproduce the array option issue you are describing, but it works fine on my machine, both when the plugin option is defined as an array and when it is defined as a scalar value (like `null` or a string) and then overwritten with an array in the site config.

If you are still having this issue, please open an issue about this with more details on how to reproduce the issue and with the result you are getting. I will then take another look at it.
